### PR TITLE
feat(shared-docs): process mermaid code blocks into styled SVG images

### PR DESCRIPTION
**Pull request by josephperrott**
Include processing of mermaid code blocks in markdown generation.

### CODE REVIEW

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [inline docs/markdown/guides/mermaid/index.ts:36]: We should resolve the `CHROME_BIN` directly and not rely on the workspace execroot assumption

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [inline docs/markdown/guides/mermaid/index.ts:28]: Ideally resolve via `runfiles` as well as this is a CJS/ESM interop we shouldn't be introducing I think

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [inline docs/markdown/guides/mermaid/index.ts:40]: Nit: Add a comment on why `bypassCSP`

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [inline docs/markdown/guides/mermaid/index.ts:12]: 👍

File: docs/markdown/BUILD.bazel
Comment (devversion) [inline docs/markdown/BUILD.bazel:113]: This has implications I think because: Chromium on Darwin has spaces in its files. Remember all the reports in Slack where macOS team memebers get errors because runfiles in bazel don't support spaces.

File: docs/markdown/BUILD.bazel
Comment (devversion) [inline docs/markdown/BUILD.bazel:116]: Are there ways to have sub-targets for guides with mermaid to only fetch the browsers when needed?

File: docs/markdown/guides/testing/BUILD.bazel
Comment (devversion) [inline docs/markdown/guides/testing/BUILD.bazel:28]: Nit: Would be good to keep that comment and add one for playwright.

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [inline docs/markdown/guides/mermaid/index.ts:2]: Why aren't we using Pupeteer?

File: docs/markdown/guides/mermaid/index.ts
Comment (jelbourn) [inline docs/markdown/guides/mermaid/index.ts:34]: How much overhead is there in launching the browser per diagram? Would it be valuable to keep the browser running until the end of this process in case the current run has multiple diagrams?

File: docs/markdown/guides/mermaid/styling.ts
Comment (jelbourn) [inline docs/markdown/guides/mermaid/styling.ts:6]: Does this CSS strictly need to be included as part of the pipeline, or could it just be included directly (perhaps as a Sass mixin) into the docs site? 

If not, could this live in a `.css` file that we import into this file?

File: docs/markdown/guides/mermaid/index.ts
Comment (josephperrott) [reply docs/markdown/guides/mermaid/index.ts:40]: I removed it, I was running into some issues at one point I think because of injecting the script tag after initial load, but it seems to work now.

File: docs/markdown/guides/mermaid/index.ts
Comment (josephperrott) [reply docs/markdown/guides/mermaid/index.ts:2]: I couldn't get puppeteer to work, I could not figure out why

File: docs/markdown/guides/mermaid/index.ts
Comment (josephperrott) [reply docs/markdown/guides/mermaid/index.ts:34]: I don't believe its that much overhead, and marked doesn't have really comprehensive hooks to use to determine when we can close the browser as far as I know.  I can put a TODO to investigate and revisit it later.

File: docs/markdown/guides/mermaid/styling.ts
Comment (josephperrott) [reply docs/markdown/guides/mermaid/styling.ts:6]: It likely could, I just ported this from the previous version.

Comment (josephperrott) [reply docs/markdown/guides/mermaid/styling.ts:6]: Done

File: docs/markdown/guides/mermaid/index.ts
Comment (josephperrott) [reply docs/markdown/guides/mermaid/index.ts]: `CHROME_BIN` comes in as `../org_chromium_chromium_linux_x64/chrome-linux/chrome` and doesn't resolve correctly in both test and build environments.  I think we end up having to do this because we are not in the run context, but the build context and bazel lays files out differently.

File: docs/markdown/BUILD.bazel
Comment (josephperrott) [reply docs/markdown/BUILD.bazel:113]: Should I switch to using firefox for it you think?

File: docs/markdown/BUILD.bazel
Comment (josephperrott) [reply docs/markdown/BUILD.bazel:116]: I suppose we could have a flag on a macro so that we note if it uses mermaid, but I am not sure if thats worth doing, thoughts?

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [reply docs/markdown/guides/mermaid/index.ts:2]: We should try to get it work as we are introducing another dependency for this and we should consistent automation, if we already have webdriver + pupeteer in all repos

Comment (devversion) [reply docs/markdown/guides/mermaid/index.ts:2]: I'm happy to help with this

File: docs/markdown/BUILD.bazel
Comment (devversion) [reply docs/markdown/BUILD.bazel:116]: If possible, would be nice. Browsers have issues the windows spaces, so that would be good, and it does require fetching the downloads

File: docs/markdown/BUILD.bazel
Comment (devversion) [reply docs/markdown/BUILD.bazel:113]: Mhmm. Interesting idea. Might be worth doing yeah 😄

File: docs/markdown/guides/mermaid/index.ts
Comment (devversion) [reply docs/markdown/guides/mermaid/index.ts]: Interesting.  That is a Bazel short path and _should_ work in practice. Using it directly as a execroot path is reasonable though. I think this is what you do? please add a comment 😄

File: docs/markdown/BUILD.bazel
Comment (josephperrott) [reply docs/markdown/BUILD.bazel:116]: Done.

File: docs/markdown/BUILD.bazel
Comment (josephperrott) [reply docs/markdown/BUILD.bazel:113]: So firefox we don't have windows support.  So I am not sure what our options really are :D

File: docs/markdown/guides/mermaid/index.ts
Comment (josephperrott) [reply docs/markdown/guides/mermaid/index.ts:2]: I think that for now I would like to unblock the work within adev on angular/angular that is blocked on getting this support in place.

We can replace this at any point as its happening under the hood anyway. As I tried and investigated it seemed mostly to be issues around the fact that its trying to create a bunch of tmp files/directories in read only workspaces.

Comment (devversion) [reply docs/markdown/guides/mermaid/index.ts:2]: I did run it with Bazel in the past to perform screenshot diffing.

Can you create an issue for tracking? We should definitely revisit what we use for automation, so that we use the same thing consistently

Comment (josephperrott) [reply docs/markdown/guides/mermaid/index.ts:2]: Done https://github.com/angular/dev-infra/issues/1784

File: docs/markdown/BUILD.bazel
Comment (josephperrott) [reply docs/markdown/BUILD.bazel:135]: In my testing I did not have any issues on os x

Comment (devversion) [reply docs/markdown/BUILD.bazel:135]: Okay, we might need to revisit this though if comes up. See all the reports for integration tests in #fw-caretaker

File: docs/markdown/guides/parse.ts
Comment (devversion) [inline docs/markdown/guides/parse.ts:58]: This should be in the `./handle-tokens` `mermaid` `if` right?

File: docs/markdown/guides/testing/bootstrap.init.ts
Comment (devversion) [inline docs/markdown/guides/testing/bootstrap.init.ts:1]: Is this what you are saying?

```suggestion
// Set HANDLE_MERMAID to false as it's not needed for unit testing
```

**Review by devversion — APPROVED** [review]

**Comment by josephperrott** [comment]
This PR was merged into the repository by commit 24bfb96956c61fe6e58c18001214acfe612704cd.

**Pull request closed without merging**
